### PR TITLE
Add parser for lambda

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/LambdaExpression.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ */
+package com.facebook.presto.sql.tree;
+
+import java.util.List;
+import java.util.Objects;
+
+public class LambdaExpression
+        extends Expression
+{
+    private final List<QualifiedName> name;
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/LambdaExpression.java]: arguments?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/LambdaExpression.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    private final Expression expression;
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/LambdaExpression.java]: rename to `body`?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/VariableReference.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.tree;
+
+public class VariableReference
+        extends Expression
+{
+    private final QualifiedName name;
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/VariableReference.java]: Why is this a QualifiedName? It should be raw identifier.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 4]
```diff
@@ -1005,6 +1007,28 @@ public void testUnnest()
     }
 
     @Test
+    public void testLambda()
+            throws Exception
+    {
+        assertStatement("SELECT x -> sin(x) FROM t",
```
Comment (martint) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: Interestingly, that wouldn't be a valid query :)

Use `assertExpression`, instead.

**Comment by martint** [comment]
Add tests to `TestAnalyzer` to ensure semantic analysis fails with reasonable errors when lambdas are used.

**Comment by haozhun** [comment]
`VariableReference` is not used. It will be generated in `TranslationMap`. But that happens after initial analysis. And I can't put initial analysis in this pull request because that depends on better type unification in #3409

**Comment by martint** [comment]
Looks good

**Pull request merged**